flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java

```diff
@@ -0,0 +1,224 @@
+                    DEFAULT_CHECK_FLUSH_PERIOD_DURATION_MS,
+                    TimeUnit.MILLISECONDS);
+        }
+
+        this.isRunning = true;
+    }
+
+    @Override
+    public void registerBufferFlushCallback(Runnable userBufferFlushCallBack) {
+        bufferFlushCallbacks.add(userBufferFlushCallBack);
+    }
+
+    /**
+     * Request a {@link BufferBuilder} instance from {@link BufferPool} for a specific owner. The
+     * {@link TieredStorageMemoryManagerImpl} will not check whether a buffer can be requested and
+     * only record the total number of requested buffers. If the buffers in the {@link BufferPool}
+     * is not enough, this will trigger each tiered storage to flush buffers as much as possible.
+     *
+     * @return the requested buffer
+     */
+    @Override
+    public BufferBuilder requestBufferBlocking() {
+        checkIsRunning();
+
+        MemorySegment requestedBuffer = null;
+        try {
+            requestedBuffer = bufferPool.requestMemorySegmentBlocking();
+        } catch (Throwable throwable) {
+            ExceptionUtils.rethrow(throwable, "Failed to request memory segments.");
+        }
+        numRequestedBuffers.incrementAndGet();
+        checkShouldFlushCachedBuffers();
+        return new BufferBuilder(checkNotNull(requestedBuffer), this::recycleBuffer);
+    }
+
+    @Override
+    public int getMaxAllowedBuffers(Object owner) {
+        checkIsRunning();
+
+        TieredStorageMemorySpec ownerMemorySpec = checkNotNull(tieredMemorySpecs.get(owner));
+        if (ownerMemorySpec.isMemoryReleasable()) {
+            return Integer.MAX_VALUE;
+        } else {
+            int ownerGuaranteedBuffers = ownerMemorySpec.getNumGuaranteedBuffers();
+            return bufferPool.getNumBuffers()
+                    - numRequestedBuffers.get()
+                    - numTotalGuaranteedBuffers
+                    + ownerGuaranteedBuffers;
+        }
+    }
+
+    @Override
+    public void release() {
+        checkState(numRequestedBuffers.get() == 0, "Leaking buffers.");
+        if (executor != null) {
+            executor.shutdown();
+            try {
+                if (!executor.awaitTermination(5L, TimeUnit.MINUTES)) {
+                    throw new TimeoutException(
+                            "Timeout for shutting down the cache flush executor.");
+                }
+            } catch (Exception e) {
+                ExceptionUtils.rethrow(e);
+            }
+        }
+    }
+
+    private void checkShouldFlushCachedBuffers() {
+        if (shouldFlushBuffers()) {
+            bufferFlushCallbacks.forEach(Runnable::run);
+        }
+    }
+
+    private boolean shouldFlushBuffers() {
+        synchronized (this) {
+            int numTotal = bufferPool.getNumBuffers();
+            int numRequested = numRequestedBuffers.get();
+            return numRequested >= numTotal
+                    || (numRequested * 1.0 / numTotal) >= numBuffersTriggerFlushRatio;
+        }
```

Comment:
This does not guarantees the atomicity, unless all access to `bufferPool` and `numRequestedBuffers` are guarded by `this`.

Reply:
FIxed.